File: src/test/java/com/vjiki/music/controller/SongLikeControllerIntegrationTest.java
package com.vjiki.music.controller;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.user;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;

import com.fasterxml.jackson.databind.JsonNode;
import com.vjiki.music.dto.SongLikeRequest;
import com.vjiki.music.dto.SongLikeResponse;
import com.vjiki.music.entity.AccessLevel;
import com.vjiki.music.entity.AuthProvider;
import com.vjiki.music.entity.Like;
import com.vjiki.music.entity.Song;
import com.vjiki.music.entity.User;
import com.vjiki.music.repository.DislikeRepository;
import com.vjiki.music.repository.LikeRepository;
import com.vjiki.music.repository.SongRepository;
import com.vjiki.music.repository.UserRepository;
import com.vjiki.music.service.SongLikeService;

@DisplayName("SongLikeController integration tests")
class SongLikeControllerIntegrationTest extends AbstractControllerIntegrationTest {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private SongRepository songRepository;

    @Autowired
    private LikeRepository likeRepository;

    @Autowired
    private DislikeRepository dislikeRepository;

    @Autowired
    private SongLikeService songLikeService;

    private User createUser(String email, String nickname) {
        return userRepository.saveAndFlush(User.builder()
                .email(email)
                .nickname(nickname)
                .provider(AuthProvider.LOCAL)
                .accessLevel(AccessLevel.USER)
                .isActive(true)
                .createdBy("system")
                .modifiedBy("system")
                .build());
    }

    private Song createSong() {
        return songRepository.saveAndFlush(Song.builder()
                .artists(Map.of("default", List.of("Artist")))
                .audioUrls(Map.of("default", "http://audio.com/song.mp3"))
                .coverUrls(Map.of("default", "http://cover.com/cover.jpg"))
                .title("Test Song")
                .active(true)
                .createdBy("system")
                .modifiedBy("system")
                .build());
    }

    @Nested
    @DisplayName("POST /api/v1/song-likes/like")
    class LikeSong {

        @Test
        @DisplayName("should like a song")
        void shouldLikeASong() throws Exception {
            User user = createUser("test@example.com", "testuser");
            Song song = createSong();

            SongLikeRequest request = new SongLikeRequest(user.getId(), song.getId());

            mockMvc.perform(post("/api/v1/song-likes/like")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(request))
                            .with(user("test").roles("USER")))
                    .andExpect(status().isOk());

            boolean isLiked = likeRepository
                    .existsByUserIdAndSongIdAndRevokedAtIsNull(user.getId(), song.getId());
            assertThat(isLiked).isTrue();
        }
    }

    @Nested
    @DisplayName("POST /api/v1/song-likes/dislike")
    class DislikeSong {

        @Test
        @DisplayName("should dislike a song")
        void shouldDislikeASong() throws Exception {
            User user = createUser("test2@example.com", "testuser2");
            Song song = createSong();

            SongLikeRequest request = new SongLikeRequest(user.getId(), song.getId());

            mockMvc.perform(post("/api/v1/song-likes/dislike")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(request))
                            .with(user("test").roles("USER")))
                    .andExpect(status().isOk());

            boolean isDisliked = dislikeRepository
                    .existsByUserIdAndSongIdAndRevokedAtIsNull(user.getId(), song.getId());
            assertThat(isDisliked).isTrue();
        }
    }

    @Nested
    @DisplayName("SongLikeService like/dislike info")
    class LikeDislikeInfo {

        @Test
        @DisplayName("should return like/dislike info")
        void shouldReturnLikeDislikeInfo() throws Exception {
            User user = createUser("test3@example.com", "testuser3");
            Song song = createSong();

            likeRepository.saveAndFlush(Like.builder()
                    .user(user)
                    .song(song)
                    .createdBy("system")
                    .build());

            SongLikeResponse info = songLikeService.getLikeDislikeInfo(user.getId(), song.getId());

            // There is no HTTP endpoint for this, so assert on the serialized wire format instead.
            JsonNode json = objectMapper.readTree(objectMapper.writeValueAsString(info));
            assertThat(json.path("isLiked").asBoolean()).isTrue();
            assertThat(json.path("isDisliked").asBoolean()).isFalse();
            assertThat(json.has("likesCount")).isTrue();
            assertThat(json.has("dislikesCount")).isTrue();
        }
    }
}

File: bin/test/com/vjiki/music/controller/SongLikeControllerIntegrationTest.kt
package com.vjiki.music.controller

import com.vjiki.music.config.TestContainersConfig
import com.vjiki.music.dto.SongLikeRequest
import com.vjiki.music.entity.*
import com.vjiki.music.repository.*
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.extensions.spring.SpringExtension
import io.kotest.matchers.shouldBe
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.user
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import org.springframework.transaction.annotation.Transactional
import com.fasterxml.jackson.databind.ObjectMapper

@SpringBootTest
@AutoConfigureMockMvc
@ContextConfiguration(classes = [TestContainersConfig::class])
@Transactional
open class SongLikeControllerIntegrationTest : DescribeSpec() {

    override fun extensions() = listOf(SpringExtension)

    @Autowired
    lateinit var mockMvc: MockMvc

    @Autowired
    lateinit var objectMapper: ObjectMapper

    @Autowired
    lateinit var userRepository: UserRepository

    @Autowired
    lateinit var songRepository: SongRepository

    @Autowired
    lateinit var likeRepository: LikeRepository

    @Autowired
    lateinit var dislikeRepository: DislikeRepository

    init {
        describe("POST /api/v1/songs/{songId}/like") {
        it("should like a song") {
            val user = userRepository.save(User(
                email = "test@example.com",
                nickname = "testuser",
                provider = AuthProvider.LOCAL,
                accessLevel = AccessLevel.USER,
                isActive = true,
                createdBy = "system",
                modifiedBy = "system"
            ))
            val song = songRepository.save(Song(
                artists = mapOf("default" to listOf("Artist")),
                audioUrls = mapOf("default" to "http://audio.com/song.mp3"),
                coverUrls = mapOf("default" to "http://cover.com/cover.jpg"),
                title = "Test Song",
                active = true,
                createdBy = "system",
                modifiedBy = "system"
            ))

            val request = SongLikeRequest(user.id, song.id)

            mockMvc.perform(
                post("/api/v1/songs/{songId}/like", song.id)
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(request))
                    .with(user("test").roles("USER"))
            )
                .andExpect(status().isOk)

            val isLiked = likeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(user.id, song.id)
            isLiked shouldBe true
        }
    }

    describe("POST /api/v1/songs/{songId}/dislike") {
        it("should dislike a song") {
            val user = userRepository.save(User(
                email = "test2@example.com",
                nickname = "testuser2",
                provider = AuthProvider.LOCAL,
                accessLevel = AccessLevel.USER,
                isActive = true,
                createdBy = "system",
                modifiedBy = "system"
            ))
            val song = songRepository.save(Song(
                artists = mapOf("default" to listOf("Artist")),
                audioUrls = mapOf("default" to "http://audio.com/song.mp3"),
                coverUrls = mapOf("default" to "http://cover.com/cover.jpg"),
                title = "Test Song",
                active = true,
                createdBy = "system",
                modifiedBy = "system"
            ))

            val request = SongLikeRequest(user.id, song.id)

            mockMvc.perform(
                post("/api/v1/songs/{songId}/dislike", song.id)
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(request))
                    .with(user("test").roles("USER"))
            )
                .andExpect(status().isOk)

            val isDisliked = dislikeRepository.existsByUserIdAndSongIdAndRevokedAtIsNull(user.id, song.id)
            isDisliked shouldBe true
        }
    }

    describe("GET /api/v1/songs/{songId}/like-info") {
        it("should return like/dislike info") {
            val user = userRepository.save(User(
                email = "test3@example.com",
                nickname = "testuser3",
                provider = AuthProvider.LOCAL,
                accessLevel = AccessLevel.USER,
                isActive = true,
                createdBy = "system",
                modifiedBy = "system"
            ))
            val song = songRepository.save(Song(
                artists = mapOf("default" to listOf("Artist")),
                audioUrls = mapOf("default" to "http://audio.com/song.mp3"),
                coverUrls = mapOf("default" to "http://cover.com/cover.jpg"),
                title = "Test Song",
                active = true,
                createdBy = "system",
                modifiedBy = "system"
            ))

            likeRepository.save(Like(
                user = user,
                song = song,
                createdBy = "system"
            ))

            mockMvc.perform(
                get("/api/v1/songs/{songId}/like-info", song.id)
                    .param("userId", user.id.toString())
                    .with(user("test").roles("USER"))
            )
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.isLiked").value(true))
                .andExpect(jsonPath("$.isDisliked").value(false))
                .andExpect(jsonPath("$.likesCount").exists())
                .andExpect(jsonPath("$.dislikesCount").exists())
        }
    }
    }
}

